[CSS] Fix wrong scoping roots with nested @scope

mdubet opened the pull request:
#### 8a947a7a91b87990311a8accea11c2acf7446439
<pre>
[CSS] Fix wrong scoping roots with nested @scope
<a href="https://bugs.webkit.org/show_bug.cgi?id=266248">https://bugs.webkit.org/show_bug.cgi?id=266248</a>
<a href="https://rdar.apple.com/119511916">rdar://119511916</a>

Reviewed by Tim Nguyen.

The @scope are stored from child to parent (innermost to outermost)
while the scoping roots are evaluated in the reverse order.
This patch uses outermost to innermost for everything.

* LayoutTests/fast/css/scope-at-rule-expected.html:
* LayoutTests/fast/css/scope-at-rule.html:
* Source/WebCore/style/RuleSet.cpp:
(WebCore::Style::RuleSet::scopeRulesFor const):

Canonical link: <a href="https://commits.webkit.org/271941@main">https://commits.webkit.org/271941@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b7e72bcd606a93f0ba1531fe16af2114c104bc6a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/30015 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/8677 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/31327 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/32517 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/27131 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/30675 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/10868 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/5929 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/27173 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/30309 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/7279 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/25598 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/6215 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/6373 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/26690 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/33854 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/27374 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/27122 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/32565 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/6311 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/4483 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/30363 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/8055 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/26491 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7132 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/7060 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/6838 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on LayoutTests/fast/css/scope-at-rule.html:
I've said this before, but we should export this to WPT at some point

mdubet replied:
Eventually we will

nt1m commented on Source/WebCore/style/RuleSet.cpp:
```diff
@@ -473,6 +473,9 @@ Vector<Ref<const StyleRuleScope>> RuleSet::scopeRulesFor(const RuleData& ruleDat
         identifier = query.parent;
     };
 
+    // Order scopes from outermost to innermost.
+    queries.reverse();
```
Can we append in reverse order directly? or is that not trivial?

mdubet replied:
We can preprend into the `WTF::Vector` (`insert(0,value)`) Not sure if it's better though

nt1m replied:
You can use a `Deque` to prepend efficiently.

mdubet committed:
[CSS] Fix wrong scoping roots with nested @scope
https://bugs.webkit.org/show_bug.cgi?id=266248
rdar://119511916

Reviewed by Tim Nguyen.

The @scope are stored from child to parent (innermost to outermost)
while the scoping roots are evaluated in the reverse order.
This patch uses outermost to innermost for everything.

* LayoutTests/fast/css/scope-at-rule-expected.html:
* LayoutTests/fast/css/scope-at-rule.html:
* Source/WebCore/style/RuleSet.cpp:
(WebCore::Style::RuleSet::scopeRulesFor const):

Canonical link: https://commits.webkit.org/271941@main
